Add stubs to keyword docs

ecstatic-morse opened the pull request:
Resolves #60779.

This commit gives each stable keyword a short entry in the "Keywords" section in the docs for `std`. The newly added entries are only a single line each and contain the main purpose of the keyword. I changed some of the existing summary lines for consistency's sake. Each line is either an imperative ("name the type of a trait object" for `dyn`), or an object ("An abstract data type" for `enum`). I tried to avoid using the keyword itself or the word "keyword" in the summary.

Later commits can flesh out each keyword with an example for each context in which it can appear as well as a link to the appropriate part of the rust book.

**edit:**
Here's the list of keywords and summaries (sans formatting) to ease reviewing. I'll try to keep this up to date as I make changes:

keyword | summary
-- | --
Self | The implementing type within a `trait` or `impl` block, or the current type within a type definition.
as | Cast between types, or rename an import.
async | ExperimentalReturn a Future instead of blocking the current thread.
await | ExperimentalSuspend execution until the result of a Future is ready.
break | Exit early from a loop.
const | Compile-time constants and deterministic functions.
continue | Skip to the next iteration of a loop.
crate | A Rust binary or library.
dyn | Name the type of a trait object.
else | What to do when an if condition does not hold.
enum | A type that can be any one of several variants.
extern | Link to or import external code.
false | A value of type bool representing logical false.
fn | A function or function pointer.
for | Iteration with in, trait implementation with impl, or higher-ranked trait bounds (for<'a>).
if | Evaluate a block if a condition holds.
impl | Implement some functionality for a type.
in | Iterate over a series of values with for.
let | Bind a value to a variable.
loop | Loop indefinitely.
match | Control flow based on pattern matching.
mod | Organize code into modules.
move | Capture a closure's environment by value.
mut | A mutable binding, reference, or pointer.
pub | Make an item visible to others.
ref | Bind by reference during pattern matching.
return | Return a value from a function.
self | The receiver of a method, or the current module.
static | A place that is valid for the duration of a program.
struct | A type that is composed of other types.
super | The parent of the current module.
trait | A common interface for a class of types.
true | A value of type bool representing logical true.
type | Define an alias for an existing type.
union | The Rust equivalent of a C-style union.
unsafe | Code or interfaces whose memory safety cannot be verified by the type system.
use | Import or rename items from other crates or modules.
where | Add constraints that must be upheld to use an item.
while | Loop while a condition is upheld.

rust-highfive commented:
r? @shepmaster

(rust_highfive has picked a reviewer for you, use r? to override)

ecstatic-morse commented:
r? @petrochenkov

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/199965753) ([raw log](https://api.travis-ci.com/v3/job/199965753/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:16db2b7e:start=1557827296584271534,finish=1557827297344107671,duration=759836137
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
$ export GCP_CACHE_BUCKET=rust-lang-ci-cache
$ export AWS_ACCESS_KEY_ID=AKIA46X5W6CZEJZ6XT55
---

[00:04:42] travis_fold:start:tidy
travis_time:start:tidy
tidy check
[00:04:42] tidy error: /checkout/src/libstd/keyword_docs.rs:860: TODO is deprecated; use FIXME
[00:04:47] some tidy checks failed
[00:04:47] 
[00:04:47] 
[00:04:47] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:04:47] 
[00:04:47] 
[00:04:47] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:04:47] Build completed unsuccessfully in 0:01:12
[00:04:47] Build completed unsuccessfully in 0:01:12
[00:04:47] make: *** [tidy] Error 1
[00:04:47] Makefile:67: recipe for target 'tidy' failed
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:1407b7a5
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Tue May 14 09:53:16 UTC 2019
---
travis_time:end:222d9e12:start=1557827597163396995,finish=1557827597169113527,duration=5716532
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0872c3b0
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:22bcaba6
travis_time:start:22bcaba6
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:071cd900
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/199973545) ([raw log](https://api.travis-ci.com/v3/job/199973545/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:179a9e56:start=1557829011799816242,finish=1557829012537681079,duration=737864837
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
$ export GCP_CACHE_BUCKET=rust-lang-ci-cache
$ export AWS_ACCESS_KEY_ID=AKIA46X5W6CZEJZ6XT55
---
travis_time:start:test_codegen
Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:19:33] 
[01:19:33] running 143 tests
[01:19:36] i..iii.....iii..iiii.....i......................i...i................i.....i..........ii.i..i..i.ii. 100/143
[01:19:37] test result: ok. 113 passed; 0 failed; 30 ignored; 0 measured; 0 filtered out
[01:19:37] 
[01:19:37]  finished in 4.608
[01:19:37] travis_fold:end:test_codegen
---
travis_time:start:test_assembly
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:19:39] 
[01:19:39] running 9 tests
[01:19:39] iiiiiiiii
[01:19:39] 
[01:19:39]  finished in 0.148
[01:19:39] travis_fold:end:test_assembly

---
travis_time:start:test_debuginfo
Check compiletest suite=debuginfo mode=debuginfo-both (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:19:55] 
[01:19:55] running 122 tests
[01:20:20] .iiiii...i.....i..i...i..i.i.i..i.ii..i.i.....i..i....i..........iiii..........i...ii...i.......ii.i 100/122
[01:20:25] .i.i......iii.i.....ii
[01:20:25] 
[01:20:25]  finished in 30.530
[01:20:25] travis_fold:end:test_debuginfo

---
[01:41:43] travis_fold:end:stage0-linkchecker

[01:41:43] travis_time:end:stage0-linkchecker:start=1557835124511247893,finish=1557835126474664071,duration=1963416178

[01:41:44] std/keyword.mod.html:1: broken link - book/ch07-02-modules-and-use-to-control-scope-and-privacy.html
[01:41:45] std/keyword.super.html:1: broken link - book/ch07-02-modules-and-use-to-control-scope-and-privacy.html
[01:41:45] std/index.html:295: broken link - book/ch07-02-modules-and-use-to-control-scope-and-privacy.html
[01:41:45] std/index.html:304: broken link - book/ch07-02-modules-and-use-to-control-scope-and-privacy.html
[01:41:51] thread 'main' panicked at 'found some broken links', src/tools/linkchecker/main.rs:41:9
[01:41:51] 
[01:41:51] 
[01:41:51] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/linkchecker" "/checkout/obj/build/x86_64-unknown-linux-gnu/doc"
[01:41:51] expected success, got: exit code: 101
[01:41:51] expected success, got: exit code: 101
[01:41:51] 
[01:41:51] 
[01:41:51] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:41:51] Build completed unsuccessfully in 0:34:00
[01:41:51] Makefile:48: recipe for target 'check' failed
[01:41:51] make: *** [check] Error 1
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:1775d9fe
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Tue May 14 11:58:55 UTC 2019
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

shepmaster commented:
/cc @rust-lang/docs

CryZe commented on src/libstd/keyword_docs.rs:
```diff
@@ -166,15 +183,15 @@ mod crate_keyword { }
 #[doc(keyword = "extern")]
 //
-/// For external connections in Rust code.
+/// Link to or import external code.
```
extern is used for specifying the ABI. So "link to or import" is not really true.

ecstatic-morse replied:
"import external code" alludes to `extern crate foo` syntax, which is still valid but pretty much useless on 2018, maybe this should be removed?

Centril replied:
Both should be kept.

CryZe commented on src/libstd/keyword_docs.rs:
```diff
@@ -283,7 +305,7 @@ mod fn_keyword { }
 #[doc(keyword = "for")]
 //
-/// The `for` keyword.
+/// Iterate over a series of values with [`in`], or implement a trait for a type with [`impl`].
```
This is missing HRTB `for<'a>`

ecstatic-morse replied:
That's true, although `for<'a>` is mentioned in the next paragraph. I tried to avoid simply enumerating all of the uses of a keyword in the summary and instead giving the reader a sense of its meaning in rust programs. That's why `impl` and `const` are so vague. I wasn't able to do this with `for`, since looping and `impl Trait for T` are fundamentally different.

ecstatic-morse replied:
The hope is that even if more uses of a keyword are added to the language, its general meaning will stay the same. For example, `impl Trait` describes what functionality a return type implements.

ecstatic-morse replied:
```suggestion
/// Iteration with [`in`], trait implementation with [`impl`], or [higher-ranked trait bounds] (`for<'a>`).
```

After taking a break and coming back, I think I'm in favor of including HRTBs in the summary line. It seems conceivable that someone would search through the `std` docs when they first encounter `for<'a>`, and seeing an example of it in the search results would be useful.

shepmaster commented:
I have general unease over the amount of duplication between this content, TRPL, and the reference.

petrochenkov commented:
My point in https://github.com/rust-lang/rust/issues/60779 is that *if* the main documentation page has a keyword list, then the list should be correct (perhaps with stubs).

Whether placing the keyword docs on that page is a good idea or not is a separate question.
I don't know. I remember I was weakly against when https://github.com/rust-lang/rust/pull/51140 first appeared, but I don't care too much, I guess.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/200111320) ([raw log](https://api.travis-ci.com/v3/job/200111320/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:11421b7b:start=1557854123005599575,finish=1557854125028189446,duration=2022589871
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
$ export GCP_CACHE_BUCKET=rust-lang-ci-cache
$ export AWS_ACCESS_KEY_ID=AKIA46X5W6CZEJZ6XT55
---
travis_time:start:test_codegen
Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:21:43] 
[01:21:43] running 143 tests
[01:21:45] i..iii.....iii..iiii.....i......................i..i.................i.....i..........ii.i..i..i.ii. 100/143
[01:21:47] test result: ok. 113 passed; 0 failed; 30 ignored; 0 measured; 0 filtered out
[01:21:47] 
[01:21:47]  finished in 4.469
[01:21:47] travis_fold:end:test_codegen
---
travis_time:start:test_assembly
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:21:49] 
[01:21:49] running 9 tests
[01:21:49] iiiiiiiii
[01:21:49] 
[01:21:49]  finished in 0.145
[01:21:49] travis_fold:end:test_assembly

---
travis_time:start:test_debuginfo
Check compiletest suite=debuginfo mode=debuginfo-both (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:22:04] 
[01:22:04] running 122 tests
[01:22:27] .iiiii...i.....i..i...i..i.i.i..i.ii..i.i.....i..i....i..........iiii..........i...ii...i.......ii.i 100/122
[01:22:32] .i.i......iii.i.....ii
[01:22:32] 
[01:22:32]  finished in 28.270
[01:22:32] travis_fold:end:test_debuginfo

---
[01:43:26] travis_fold:end:stage0-linkchecker

[01:43:26] travis_time:end:stage0-linkchecker:start=1557860339911112823,finish=1557860341993826775,duration=2082713952

[01:43:28] std/keyword.mod.html:1: absolute path - /nightly/book/ch07-02-defining-modules-to-control-scope-and-privacy.html
[01:43:28] std/keyword.super.html:1: absolute path - /nightly/book/ch07-02-defining-modules-to-control-scope-and-privacy.html
[01:43:28] std/index.html:295: absolute path - /nightly/book/ch07-02-defining-modules-to-control-scope-and-privacy.html
[01:43:28] std/index.html:304: absolute path - /nightly/book/ch07-02-defining-modules-to-control-scope-and-privacy.html
[01:43:34] thread 'main' panicked at 'found some broken links', src/tools/linkchecker/main.rs:41:9
[01:43:34] 
[01:43:34] 
[01:43:34] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/linkchecker" "/checkout/obj/build/x86_64-unknown-linux-gnu/doc"
[01:43:34] expected success, got: exit code: 101
[01:43:34] expected success, got: exit code: 101
[01:43:34] 
[01:43:34] 
[01:43:34] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:43:34] Build completed unsuccessfully in 0:33:18
[01:43:34] Makefile:48: recipe for target 'check' failed
[01:43:34] make: *** [check] Error 1
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:0c61f24f
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Tue May 14 18:59:10 UTC 2019
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/200147750) ([raw log](https://api.travis-ci.com/v3/job/200147750/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:15a82df0:start=1557862127582414182,finish=1557862129782117609,duration=2199703427
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
$ export GCP_CACHE_BUCKET=rust-lang-ci-cache
$ export AWS_ACCESS_KEY_ID=AKIA46X5W6CZEJZ6XT55
---
travis_time:start:test_codegen
Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:21:07] 
[01:21:07] running 143 tests
[01:21:10] i..iii.....iii..iiii.....i......................i...i................i.....i..........ii.i..i..i.ii. 100/143
[01:21:12] test result: ok. 113 passed; 0 failed; 30 ignored; 0 measured; 0 filtered out
[01:21:12] 
[01:21:12]  finished in 4.666
[01:21:12] travis_fold:end:test_codegen
---
travis_time:start:test_assembly
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:21:14] 
[01:21:14] running 9 tests
[01:21:14] iiiiiiiii
[01:21:14] 
[01:21:14]  finished in 0.160
[01:21:14] travis_fold:end:test_assembly

---
travis_time:start:test_debuginfo
Check compiletest suite=debuginfo mode=debuginfo-both (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:21:30] 
[01:21:30] running 122 tests
[01:21:55] .iiiii...i.....i..i...i..i.i.i..i.ii..i.i.....i..i....i..........iiii..........i...ii...i.......ii.i 100/122
[01:22:01] .i.i......iii.i.....ii
[01:22:01] 
[01:22:01]  finished in 30.903
[01:22:01] travis_fold:end:test_debuginfo

---
[01:43:41] travis_fold:end:stage0-linkchecker

[01:43:41] travis_time:end:stage0-linkchecker:start=1557868360863827443,finish=1557868362967510108,duration=2103682665

[01:43:43] std/keyword.mod.html:1: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/nightly/book/ch07-02-defining-modules-to-control-scope-and-privacy.html
[01:43:43] std/keyword.super.html:1: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/nightly/book/ch07-02-defining-modules-to-control-scope-and-privacy.html
[01:43:44] std/index.html:295: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/nightly/book/ch07-02-defining-modules-to-control-scope-and-privacy.html
[01:43:44] std/index.html:304: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/nightly/book/ch07-02-defining-modules-to-control-scope-and-privacy.html
[01:43:49] thread 'main' panicked at 'found some broken links', src/tools/linkchecker/main.rs:41:9
[01:43:49] 
[01:43:49] 
[01:43:49] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/linkchecker" "/checkout/obj/build/x86_64-unknown-linux-gnu/doc"
[01:43:49] expected success, got: exit code: 101
[01:43:49] expected success, got: exit code: 101
[01:43:49] 
[01:43:49] 
[01:43:49] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:43:49] Build completed unsuccessfully in 0:34:42
[01:43:49] make: *** [check] Error 1
[01:43:49] Makefile:48: recipe for target 'check' failed
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:058a62ba
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Tue May 14 21:12:51 UTC 2019
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -29,9 +29,14 @@
 #[doc(keyword = "const")]
 //
-/// The keyword for defining constants.
+/// Compute something at compile-time.
```
Not true since `const fn` got stable.

ecstatic-morse replied:
I considered "Assert that something can be computed at compile-time" which applies to both `const` declarations and `const fn`s and is technically more correct but also more verbose.

Centril replied:
This is not the full story of `const fn` and is in my view quite misleading. A `const fn` is a deterministic function that *can* be executed at CTFE if given CT-evaluable arguments. A `const fn` can also be evaluated at runtime.

petrochenkov replied:
It's probably not a good idea to put the full story into a single-sentence title.
"Compute something at compile-time" seems both close and vague enough.

Centril replied:
If you want to be vague, I'd at least throw in a "can" here since a `const fn` need not be evaluated at CT.

Centril replied:
My preference would be for something like:

```suggestion
/// Compile time constants and deterministic functions.
```

I think conflating `const fn` with `const` spreads a misconception I'm not too happy about.

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -116,14 +126,21 @@ mod const_keyword { }
 
+#[doc(keyword = "else")]
+//
+/// What to do when an [`if`] condition is false.
```
```suggestion
/// What to do when an [`if`] condition does not hold.
```
In lieu of https://github.com/rust-lang/rust/issues/53667.

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -29,9 +29,14 @@
 
+#[doc(keyword = "break")]
+//
+/// Exit early from a loop.
```
Not true with https://github.com/rust-lang/rust/issues/48594 but it's unstable.

GuillaumeGomez replied:
I'm not sure to understand your @Centril

Centril replied:
@GuillaumeGomez That this is not a full description of what `break` does.

Centril replied:
Alright, since we are going for "mostly right but not full story", let's call this resolved.

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -116,14 +126,21 @@ mod const_keyword { }
 #[doc(keyword = "enum")]
 //
-/// For defining enumerations.
+/// An [Algebraic Data Type][ADT].
```
A `struct` is also an ADT; `enum` is specifically a *sum type*.

ecstatic-morse replied:
True. This needs to get changed. FWIW, I think it would be better to use  "composite" and "enumerated" in the summary line as opposed to "product" and "sum", since the audience for these docs is unlikely to be familiar with them (although "composite" and "enumerated" aren't much clearer).

Centril replied:
I think "composite" isn't that familiar either. At least "product" is a bit familiar for those who have heard of a Cartesian product. However, since the goal here is to give an intuitive explanation, we can avoid both terminologies altogether:
```suggestion
/// A type that can be one of several variants.
```
or alternatively:
```suggestion
/// A type encoding the notion of choice.
```

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -360,7 +383,7 @@ mod for_keyword { }
 #[doc(keyword = "if")]
 //
-/// If statements and expressions.
+/// Traditional control flow based on boolean conditions.
```
Stops being true with https://github.com/rust-lang/rust/issues/53667 and is already false with `if let`.

ecstatic-morse replied:
True, this should probably just stay the way it is. I was trying to differentiate it from `match`.

Centril replied:
It's a good idea to differentiate, but perhaps say:
```suggestion
/// Evaluate a block if the condition holds or possibly a different one otherwise.
```

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -495,9 +518,16 @@ mod if_keyword { }
 #[doc(keyword = "let")]
 //
-/// The variable binding keyword.
+/// Bind a value to a variable.
```
`let` is used for general pattern matching, which may include binding variables but may also not.

ecstatic-morse replied:
I would say that binding a value to (one or more) variable(s) is by far the most common of usage of `let`. For bare `let` I can think of `let _ = ...` and irrefutable patterns like `let UnitStruct = v` as exceptions. `let` without binding is more useful in situations like `if let Enum::UnitVariant = v` but is pretty rare in my personal experience.

Centril replied:
Well you can say something like `let Wrapping(my_binding) = expr;` which both does irrefutable pattern matching and binds a value. Pattern matching and "irrefutable" are also terminologies we already use.

Possibly, you could go into the common case in the first line and then talk about the general case in later paragraphs or invert the order. However, I think the info about the general case should be included somehow.

ecstatic-morse replied:
How about "Bind to a value by pattern matching"?

Centril replied:
In that case:
```suggestion
/// Bind values by pattern matching.
```
(note the plural)

ecstatic-morse replied:
The right hand side of a let statement is an expression, the result of which is a single value. That value can be a composite (e.g. `(1, 2)`),  but I believe a `let` statement only operates on one value at a time. I could agree with "create zero or more bindings to a value by pattern matching" but I think it's more verbose than necessary.

Centril replied:
I see what you're saying... Upon re-reading the original, I think it's servicable for now so I'll resolve this.

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -603,9 +633,58 @@ mod let_keyword { }
+
+#[doc(keyword = "mut")]
+//
+/// A mutable binding, reference or pointer.
```
```suggestion
/// A mutable binding, reference, or pointer.
```

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -603,9 +633,58 @@ mod let_keyword { }
+
+#[doc(keyword = "pub")]
+//
+/// Make an item visible to others.
```
This isn't restricted to items, e.g. fields may also be `pub`.

ecstatic-morse replied:
Is there a better word than "something"? I know "item" has a specific meaning in Rust's grammar.

Centril replied:
cc @petrochenkov

petrochenkov replied:
Seems ok for a summary.

Centril replied:
Fair enough.

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -603,9 +633,58 @@ mod let_keyword { }
+
+#[doc(keyword = "static")]
+//
+/// Something which is valid for the duration of a program.
```
```suggestion
/// A place which is valid for the duration of a program.
```

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -603,9 +633,58 @@ mod let_keyword { }
 #[doc(keyword = "struct")]
 //
-/// The keyword used to define structs.
+/// A composite data type.
```
```suggestion
/// A nominal product type.
```

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -710,3 +789,79 @@ mod loop_keyword { }
+
+#[doc(keyword = "trait")]
+//
+/// A single interface shared across types.
```
```suggestion
/// A common interface for a class of types.
```

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -710,3 +789,79 @@ mod loop_keyword { }
+
+#[doc(keyword = "type")]
+//
+/// Give a new name to an existing type.
```
Not really true for associated types.

ecstatic-morse replied:
There are four uses that I'm aware of: `type` aliases, declaring an associated type in a trait definition, defining an associated type in a trait implementation, and `extern type`. I think that this statement also applies when defining an associated type (I give a type the name `<T as Trait>::AssocType`) and declaring an `extern type` (I give some external type a name, but don't specify its layout). It doesn't really apply when defining a trait with an associated type. 

Perhaps this is good enough?

Centril replied:
Yeah fair enough, but it would be good to elaborate more later.

At your option, here's one possible change:
```suggestion
/// Define an alias for an existing type.
```

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -710,3 +789,79 @@ mod loop_keyword { }
+
+#[doc(keyword = "unsafe")]
+//
+/// Opt out of statically enforced memory safety.
```
Only true for `unsafe { .. }` and `unsafe impl` but not of `unsafe fn` and `unsafe trait`.
It's better to talk about that `unsafe { ... }` is a "trust me" construct and that `unsafe fn` has proof obligations.

ecstatic-morse replied:
I don't think that the distinction between "unsafe to use" and "unsafe to implement" needs to be mentioned here, although the imperative does make it sound like I'm talking exclusively about `unsafe` blocks. This line was adapted from the chapter summary in the Rust book:

> All the code we’ve discussed so far has had Rust’s memory safety guarantees enforced at compile time. However, Rust has a second language hidden inside it that doesn’t enforce these memory safety guarantees: it’s called  unsafe Rust...

Centril replied:
> I don't think that the distinction between "unsafe to use" and "unsafe to implement" needs to be mentioned here,

That's not the distinction I'm going for since your current change covers both `unsafe impl` and `unsafe { ... }`. Rather, it is the distinction between `unsafe { ... }` and `unsafe fn`.

I think I might go for something more akin to:
```suggestion
/// Verify safety yourself or define an unsafe abstraction.
```

The point here is to tell the reader that you can't just do whatever you want in `unsafe { ... }` and that certain invariants the compiler expects still must be upheld.

ecstatic-morse replied:
Another option, once again paraphrased from the book, but split into "interfaces" and "code" for distinguishing `unsafe fn`/`unsafe trait` versus `unsafe { ... }`

```suggestion
/// Code or interfaces whose memory safety cannot be guaranteed at compile time.
```

Centril replied:
I think `Verify` is more imperative, but I'd change your version to:
```suggestion
/// Code or interfaces whose safety cannot be verified by the type system.
```

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -710,3 +789,79 @@ mod loop_keyword { }
+
+#[doc(keyword = "where")]
+//
+/// Add bounds to a type parameter.
```
```suggestion
/// Add constraints that must be upheld to use the item.
```

Not restricted to type parameters.

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -710,3 +789,79 @@ mod loop_keyword { }
+
+#[doc(keyword = "while")]
+//
+/// Loop while a condition remains true.
```
```suggestion
/// Loop while a condition is upheld.
```
In lieu of https://github.com/rust-lang/rust/issues/53667.

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -710,3 +789,79 @@ mod loop_keyword { }
+
+#[doc(keyword = "union")]
+//
+/// An unsafe [Algebraic Data Type][ADT] equivalent to a C-style `union`.
```
In what way is a union an ADT?

ecstatic-morse replied:
It's not, my mistake. See the comments on `enum`.

petrochenkov commented:
By using literal stubs instead of summaries you could both resolve https://github.com/rust-lang/rust/issues/60779 and avoid @Centril's comments :)
But I guess it's to late to turn back now.

petrochenkov commented on src/libstd/keyword_docs.rs:
```diff
@@ -603,9 +633,58 @@ mod let_keyword { }
+
+#[doc(keyword = "self")]
+//
+/// The receiver of a method.
```
```suggestion
/// The receiver of a method or the current module.
```

petrochenkov commented:
r=me once @Centril's concerns are resolved.

r? @Centril

petrochenkov commented:
It would be good to add a paragraph explicitly saying that the doc is not yet written and inviting the reader to make a PR for the keywords with missing documentation body.

ecstatic-morse commented:
@petrochenkov These summary lines are visible on `doc.rust-lang.org/stable/std`, so they should at least have a uniform style. I considered:
   1. Making all keyword summary lines empty (not sure if it's possible to have no summary line but still have documentation).
   2. Leaving the existing summary lines alone and adding empty ones (looks ugly, not uniform).
   3. Using "The `x` keyword" for all keywords (uninformative and repetitive).
   4. Writing new summary lines in the style of `std` and getting the bikeshedding over with.

The last one seemed least bad.

petrochenkov commented:
>The last one seemed least bad.

It's not even less bad, it's certainly better in the long run!

ecstatic-morse commented:
@petrochenkov Will do. How about:

"The documentation for this keyword is [not yet complete](https://github.com/rust-lang/rust/issues/34601). Pull requests welcome!"

Centril commented:
Seems good to me.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/200197871) ([raw log](https://api.travis-ci.com/v3/job/200197871/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:1c24a794:start=1557872641057319656,finish=1557872644671779500,duration=3614459844
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
$ export GCP_CACHE_BUCKET=rust-lang-ci-cache
$ export AWS_ACCESS_KEY_ID=AKIA46X5W6CZEJZ6XT55
---
travis_time:start:test_codegen
Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:21:30] 
[01:21:30] running 143 tests
[01:21:33] i..iii.....iii..iiii.....i......................i...i................i......i.........ii.i..i..i.ii. 100/143
[01:21:35] test result: ok. 113 passed; 0 failed; 30 ignored; 0 measured; 0 filtered out
[01:21:35] 
[01:21:35]  finished in 4.988
[01:21:35] travis_fold:end:test_codegen
---
travis_time:start:test_assembly
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:21:38] 
[01:21:38] running 9 tests
[01:21:38] iiiiiiiii
[01:21:38] 
[01:21:38]  finished in 0.161
[01:21:38] travis_fold:end:test_assembly

---
travis_time:start:test_debuginfo
Check compiletest suite=debuginfo mode=debuginfo-both (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:21:53] 
[01:21:53] running 122 tests
[01:22:18] .iiiii...i.....i..i...i..i.i.i..i.ii..i.i.....i..i....i..........iiii..........i...ii...i.......ii.i 100/122
[01:22:22] .i.i......iii.i.....ii
[01:22:22] 
[01:22:22]  finished in 29.344
[01:22:22] travis_fold:end:test_debuginfo

---
[01:43:44] travis_fold:end:stage0-linkchecker

[01:43:44] travis_time:end:stage0-linkchecker:start=1557878877600547153,finish=1557878879729506202,duration=2128959049

[01:43:46] std/keyword.mod.html:1: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/nightly/book/ch07-02-defining-modules-to-control-scope-and-privacy.html
[01:43:46] std/keyword.super.html:1: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/nightly/book/ch07-02-defining-modules-to-control-scope-and-privacy.html
[01:43:47] std/index.html:295: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/nightly/book/ch07-02-defining-modules-to-control-scope-and-privacy.html
[01:43:47] std/index.html:304: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/nightly/book/ch07-02-defining-modules-to-control-scope-and-privacy.html
[01:43:52] thread 'main' panicked at 'found some broken links', src/tools/linkchecker/main.rs:41:9
[01:43:52] 
[01:43:52] 
[01:43:52] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/linkchecker" "/checkout/obj/build/x86_64-unknown-linux-gnu/doc"
[01:43:52] expected success, got: exit code: 101
[01:43:52] expected success, got: exit code: 101
[01:43:52] 
[01:43:52] 
[01:43:52] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:43:52] Build completed unsuccessfully in 0:34:15
[01:43:52] Makefile:48: recipe for target 'check' failed
[01:43:52] make: *** [check] Error 1
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:0dbfaa88
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Wed May 15 00:08:08 UTC 2019
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -603,9 +635,58 @@ mod let_keyword { }
+
+#[doc(keyword = "static")]
+//
+/// A place which is valid for the duration of a program.
```
```suggestion
/// A place that is valid for the duration of a program.
```
Grammar nit I found on closer inspection. :)

Centril commented:
Btw, when you're done with changes altogether, can you squash things into a single commit? (also please avoid @-mentioning folks in commits cause it generates many emails ^^).

ecstatic-morse commented:
I added `Self` to the list of keywords, added a note about the incomplete docs and a link to #34601, squashed, resolved most of the ongoing discussions, and reduced noise in Centril's inbox.

Centril commented on src/libstd/keyword_docs.rs:
```diff
@@ -603,9 +652,101 @@ mod let_keyword { }
+
+#[doc(keyword = "Self")]
+//
+/// The implementing type within traits and implementations.
```
```suggestion
/// The implementing type within traits, implementations, and type definitions.
```

ecstatic-morse replied:
Agreed. Btw, this is another one [direct from the reference](https://doc.rust-lang.org/reference/paths.html#self-1)

Centril replied:
The reference is wrong cause I never got around to fixing it after https://github.com/rust-lang/rust/issues/49303 :)

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/200232107) ([raw log](https://api.travis-ci.com/v3/job/200232107/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:087ab5dc:start=1557885636902207577,finish=1557885725018605259,duration=88116397682
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
$ export GCP_CACHE_BUCKET=rust-lang-ci-cache
$ export AWS_ACCESS_KEY_ID=AKIA46X5W6CZEJZ6XT55
---
travis_time:start:test_codegen
Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:18:49] 
[01:18:49] running 143 tests
[01:18:52] i..iii.....iii..iiii.....i......................i..i.................i.....i..........ii.i..i..i.ii. 100/143
[01:18:54] test result: ok. 113 passed; 0 failed; 30 ignored; 0 measured; 0 filtered out
[01:18:54] 
[01:18:54]  finished in 4.594
[01:18:54] travis_fold:end:test_codegen
---
travis_time:start:test_assembly
Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:18:56] 
[01:18:56] running 9 tests
[01:18:56] iiiiiiiii
[01:18:56] 
[01:18:56]  finished in 0.154
[01:18:56] travis_fold:end:test_assembly

---
travis_time:start:test_debuginfo
Check compiletest suite=debuginfo mode=debuginfo-both (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:19:11] 
[01:19:11] running 122 tests
[01:19:36] .iiiii...i.....i..i...i..i.i.i..i.ii..i.i.....i..i....i..........iiii..........i...ii...i.......ii.i 100/122
[01:19:41] .i.i......iii.i.....ii
[01:19:41] 
[01:19:41]  finished in 29.659
[01:19:41] travis_fold:end:test_debuginfo

---
[01:41:04] travis_fold:end:stage0-linkchecker

[01:41:04] travis_time:end:stage0-linkchecker:start=1557891796973844216,finish=1557891798993551479,duration=2019707263

[01:41:07] std/keyword.union.html:1: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/reference/items/unions.html
[01:41:07] std/index.html:309: broken link - /checkout/obj/build/x86_64-unknown-linux-gnu/reference/items/unions.html
[01:41:12] thread 'main' panicked at 'found some broken links', src/tools/linkchecker/main.rs:41:9
[01:41:12] 
[01:41:12] 
[01:41:12] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/linkchecker" "/checkout/obj/build/x86_64-unknown-linux-gnu/doc"
[01:41:12] expected success, got: exit code: 101
[01:41:12] expected success, got: exit code: 101
[01:41:12] 
[01:41:12] 
[01:41:12] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:41:12] Build completed unsuccessfully in 0:34:01
[01:41:12] Makefile:48: recipe for target 'check' failed
[01:41:12] make: *** [check] Error 1
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:00078169
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Wed May 15 03:43:27 UTC 2019
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

ecstatic-morse commented:
I fixed the broken doc link and noted that `Self` can also refer to "the current type within a type definition".

Centril commented:
r=me with green travis :)

Centril commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 851be33f2ab87a4088dcaf9007bdb20d3370a45e has been approved by `Centril`

<!-- @bors r=Centril 851be33f2ab87a4088dcaf9007bdb20d3370a45e -->

bors commented:
:hourglass: Testing commit 851be33f2ab87a4088dcaf9007bdb20d3370a45e with merge c2e49bf1a279ec704bfe39fa149b64783454d72d...
<!-- homu: {"type":"BuildStarted","head_sha":"851be33f2ab87a4088dcaf9007bdb20d3370a45e","merge_sha":"c2e49bf1a279ec704bfe39fa149b64783454d72d"} -->

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/rust-lang/rust/builds/112171792), [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/24610242)
Approved by: Centril
Pushing c2e49bf1a279ec704bfe39fa149b64783454d72d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Centril","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang/rust/builds/24610242","checks-travis":"https://travis-ci.com/rust-lang/rust/builds/112171792"},"merge_sha":"c2e49bf1a279ec704bfe39fa149b64783454d72d"} -->
